### PR TITLE
Reduce uncached HOC unit overview queries

### PR DESCRIPTION
**Pull request by davidsbailey**
Continues https://codedotorg.atlassian.net/browse/PLAT-1181. Follows https://github.com/code-dot-org/code-dot-org/pull/42737. This is my second and final pass at reducing query counts when loading the unit overview page for an uncached HOC unit.

## Testing story

I have reasonably high confidence in existing test coverage, given that it caught several bugs on lines I modified.

### CODE REVIEW

File: dashboard/test/integration/db_query_test.rb [hunk 1]
```diff
@@ -109,20 +109,20 @@ def setup
     student.assign_script(script)
     sign_in student
 
-    assert_cached_queries(10) do
+    assert_cached_queries(6) do
```
Comment (davidsbailey) [inline dashboard/test/integration/db_query_test.rb:112]: before: [hoc-unit-queries-10.txt](https://github.com/code-dot-org/code-dot-org/files/7280021/hoc-unit-queries-10.txt)
after: [hoc-unit-queries-06.txt](https://github.com/code-dot-org/code-dot-org/files/7280022/hoc-unit-queries-06.txt)

**Review by dmcavoy — APPROVED** [review]
Nice work Dave!

File: dashboard/app/controllers/api_controller.rb [hunk 2]
```diff
@@ -398,7 +398,7 @@ def user_progress
     if current_user
       script = Script.get_from_cache(params[:script])
       user = params[:user_id].present? ? User.find(params[:user_id]) : current_user
-      teacher_viewing_student = current_user.students.include?(user)
+      teacher_viewing_student = !current_user.student? && current_user.students.include?(user)
```
Comment (jamescodeorg) [inline dashboard/app/controllers/api_controller.rb:401]: (minor) Mostly curious, would using `current_user.teacher?` be equivalent to `!current_user.student?` here?  That seems a bit more natural to me.

Comment (davidsbailey) [reply dashboard/app/controllers/api_controller.rb:401]: yes, at the moment. However, I believe @dmcavoy is slated to start working on some of the instructor vs instructee stuff, which made this seem like a safer way to leave things.

**Review by jamescodeorg — APPROVED** [review]

**Pull request merged**